Core/Warden: More Warden improvements

Treeston opened the pull request:
* Make `LUA_STR_CHECK` functional and (somewhat more) useful by turning it into `LUA_EVAL_CHECK`
* Re-structure and cleanup how we handle Warden scheduling
* Three new checks, one of which should catch naive Lua unlockers, two of which should catch PQR (ref #23035)

_(Special thanks to @scizzydo for Warden insights.)_

Treeston committed:
More Warden improvements

Warpten commented:
No.

Shauren commented on sql/updates/world/3.3.5/9999_99_99_99_world_WARDEN.sql:
```diff
@@ -0,0 +1,17 @@
+DELETE FROM `warden_checks` WHERE `id` IN (786,787,788);
+INSERT INTO `warden_checks` (`id`,`type`,`str`,`address`,`length`,`result`, `comment`) VALUES
+(786, 139, 'forceinsecure() return issecure()', NULL, NULL, NULL, 'Detects naive Lua unlockers'),
+(787, 243, NULL, CONV('819210',16,10), 16, UNHEX('558BEC518305A013D40001A19C13D400'), 'FrameScript::Execute'),
```
you know you can just write 0x819210 and mysql will insert it properly?

Warpten replied:
for binary i personally do `x'CAFEBABE'` and it just works

Warpten commented on src/server/game/Warden/WardenWin.cpp:
```diff
@@ -413,26 +421,11 @@ void WardenWin::HandleCheckResult(ByteBuffer &buff)
             {
-                uint8 Lua_Result;
-                buff >> Lua_Result;
+                buff.read<uint8>();
```
you need to read_skip the rest if != 0

Treeston replied:
No, in fact current Warden is broken, and adding a `LUA_STR_CHECK` to the DB in current Warden will corrupt data after it.

Result should only be read if `Lua_Result` is zero, not nonzero.

Warpten replied:
i mean yeah i know the check is broken as far as its implementation in module is concerned but then you still have data if = 0

Treeston replied:
I have only ever observed result 1 for `::Execute` substitution, I believe it to be the return value from one of the lua library functions.

Anyway, added `read_skip`.

Warpten commented on src/server/game/Warden/WardenCheckMgr.h:
```diff
@@ -31,32 +32,74 @@ enum WardenActions : uint8
-    PROC_CHECK = 0x7E, // 126: uint Seed + byte[20] SHA1 + byte moluleNameIndex + byte procNameIndex + uint Offset + byte Len (check to ensure proc isn't detoured)
-    MODULE_CHECK = 0xD9  // 217: uint Seed + byte[20] SHA1 (check to ensure module isn't injected)
+    NONE_CHECK     =   0, // SKIP
+    TIMING_CHECK   =  87, // nyi
```
why nyi, it's always sent (don't ask me why). it's an empty payload for the request

Treeston replied:
the result is not processed

Warpten replied:
it is (or was)

Treeston replied:
The actual timing result is never used, just logged.

Warpten commented on src/server/game/Warden/WardenCheckMgr.h:
```diff
@@ -31,32 +32,74 @@ enum WardenActions : uint8
+    }
+}
+
+inline constexpr WorldIntConfigs GetWardenCategoryCountConfig(WardenCheckCategory category)
```
since c++17 inline is implicitely specified

Warpten commented on src/server/game/Warden/WardenCheckMgr.h:
```diff
@@ -31,32 +32,74 @@ enum WardenActions : uint8
+    MEM_CHECK      = 243, // retrieve specific memory
 };
 
+inline constexpr WardenCheckCategory GetWardenCheckCategory(WardenCheckType type)
```
same as L77

jackpoz committed:
Update .travis.yml

jackpoz committed:
Update .travis.yml

Treeston commented:
Split off a number of unrelated refactors into 5b4288f, a3971ca, fc32c3f.

Rebased this PR onto the result. This gives us a significantly more manageable changeset to review.

Treeston committed:
some refactors

Treeston committed:
ready to merge

Treeston committed:
uh that was for testing
